netenv/addresses.go

```diff
@@ -70,18 +73,33 @@ func IsMyIP(ip net.IP) (yes bool, err error) {
 	myNetworksLock.Lock()
 	defer myNetworksLock.Unlock()
 
-	// Check if the network changed.
-	if myNetworksNetworkChangedFlag.IsSet() {
-		// Reset changed flag.
-		myNetworksNetworkChangedFlag.Refresh()
-	} else if mine, matched := checkIfMyIP(ip); matched {
-		// If the network did not change, check for match immediately.
+	// Check if current data matches IP.
+	// Matching on somewhat older data is not a problem, as these IPs would not
+	// just randomly pop up somewhere else that fast.
+	mine, matched := checkIfMyIP(ip)
+	if matched {
 		return mine, nil
 	}
 
+	// Check if the network changed.
+	if !myNetworksNetworkChangedFlag.IsSet() {
+		// Network did not change, return "no match".
+		return false, nil
+	}
+
+	// Check if there was a recent error on the previous refresh.
+	if myNetworksRefreshError != nil && time.Now().Before(myNetworksRefreshFailingUntil) {
+		return false, fmt.Errorf("failed to previously refresh interface addresses: %s", myNetworksRefreshError)
+	}
+
 	// Refresh assigned networks.
 	interfaceNetworks, err := net.InterfaceAddrs()
 	if err != nil {
+		// Save error for one second.
+		// In some cases the system blocks on this call, which piles up to
+		// literally over thousand goroutines wanting to try this again.
+		myNetworksRefreshError = err
+		myNetworksRefreshFailingUntil = time.Now().Add(1 * time.Second)
 		return false, fmt.Errorf("failed to refresh interface addresses: %s", err)
```

Comment:
We could also use the "singleflight" package here and use the network interface name as the key. Otherwise LGTM